model/src/id.rs

```diff
@@ -492,630 +329,444 @@ impl GuildId {
+            }
 
-impl From<NonZeroU64> for MessageId {
-    fn from(id: NonZeroU64) -> Self {
-        MessageId(id)
-    }
-}
+            fn visit_newtype_struct<D: Deserializer<'de>>(
+                self,
+                deserializer: D,
+            ) -> Result<Self::Value, D::Error> {
+                deserializer.deserialize_any(IdVisitor::new())
+            }
 
-/// Unique ID of a role.
-#[allow(clippy::unsafe_derive_deserialize)]
-#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
-pub struct RoleId(#[serde(with = "string")] pub NonZeroU64);
+            fn visit_str<E: DeError>(self, v: &str) -> Result<Self::Value, E> {
+                let value = v.parse().map_err(|_| {
+                    let unexpected = Unexpected::Str(v);
 
-impl RoleId {
-    /// Create a non-zero role ID without checking the value.
-    ///
-    /// Equivalent to [`NonZeroU64::new_unchecked`].
-    ///
-    /// # Safety
-    ///
-    /// The value must not be zero.
-    #[allow(unsafe_code)]
-    pub const unsafe fn new_unchecked(n: u64) -> Self {
-        Self(NonZeroU64::new_unchecked(n))
-    }
+                    DeError::invalid_value(unexpected, &"non zero u64 string")
+                })?;
 
-    /// Create a non-zero role ID if the given value is not zero.
-    ///
-    /// Equivalent to [`NonZeroU64::new`].
-    pub const fn new(n: u64) -> Option<Self> {
-        #[allow(clippy::option_if_let_else)]
-        if let Some(n) = NonZeroU64::new(n) {
-            Some(Self(n))
-        } else {
-            None
+                self.visit_u64(value)
+            }
         }
-    }
 
-    /// Return the inner primitive value.
-    ///
-    /// Equivalent to [`NonZeroU64::get`].
-    pub const fn get(self) -> u64 {
-        self.0.get()
+        deserializer.deserialize_any(IdVisitor::new())
     }
 }
 
-impl Display for RoleId {
+impl<T> Display for Id<T> {
     fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
-        Display::fmt(&self.0, f)
+        Display::fmt(&self.value.get(), f)
     }
 }
 
-impl From<NonZeroU64> for RoleId {
+impl<T> From<NonZeroU64> for Id<T> {
     fn from(id: NonZeroU64) -> Self {
-        RoleId(id)
+        Self::from_nonzero(id)
     }
 }
 
-/// Unique ID of a stage.
-#[allow(clippy::unsafe_derive_deserialize)]
-#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
-pub struct StageId(#[serde(with = "string")] pub NonZeroU64);
+impl<T> FromStr for Id<T> {
+    type Err = ParseIntError;
 
-impl StageId {
-    /// Create a non-zero stage ID without checking the value.
-    ///
-    /// Equivalent to [`NonZeroU64::new_unchecked`].
-    ///
-    /// # Safety
-    ///
-    /// The value must not be zero.
-    #[allow(unsafe_code)]
-    pub const unsafe fn new_unchecked(n: u64) -> Self {
-        Self(NonZeroU64::new_unchecked(n))
-    }
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let nonzero = NonZeroU64::from_str(s)?;
```

Comment:
the diff algorithm produced pretty bad results so you'll have to remove the return line at end of this function as well
```suggestion
        NonZeroU64::from_str(s).map(Self::from_nonzero)
```